request-manager/graph/grapher.go

```diff
@@ -61,58 +61,103 @@ func (gr *Grapher) CheckSequences() (seqGraphs map[string]*Graph, seqErrors map[
-					multiple = "s"
+	// Keep checking sequences whose dependencies are satisfied until we can't
+	// anymore, either because we've checked them all, or because impossible
+	// dependencies exist.
+	seqsChecked := map[string]bool{}
+	seqsToCheck := map[string]*spec.Sequence{}
+	for k, v := range gr.sequenceSpecs {
+		seqsToCheck[k] = v
+	}
+
+	for len(seqsToCheck) != 0 {
+		newSeqChecked := false
+
+		for seqName, seqSpec := range seqsToCheck {
+			if haveAllDeps(seqsChecked, subsequences[seqName]) {
+				newSeqChecked = true
+				delete(seqsToCheck, seqName)
+				seqsChecked[seqName] = true
+
+				// Only perform sets check if this sequence built
+				// and all its subsequences have passed all checks.
+				if _, ok := seqErrors[seqName]; ok {
+					continue
+				}
+				failed := getFailedSubsequences(subsequences[seqName], seqErrors)
+				if len(failed) != 0 {
+					multiple := ""
+					if len(failed) > 1 {
+						multiple = "s"
+					}
+					seqErrors[seqName] = fmt.Errorf("subsequence%s failed checks: %s", multiple, strings.Join(failed, ", "))
+					continue
 				}
-				seqErrors[seqName] = fmt.Errorf("subsequence%s failed to build: %s", multiple, strings.Join(failed, ", "))
-				delete(seqGraphs, seqName)
-				continue
-			}
 
-			// Compare declared `sets` with what the node
-			// actually sets
-			sets := getDeclaredSets(subsequences, seqSets)
-			missing := getMissingSets(sets, nodeSpec.Sets)
-			if len(missing) != 0 {
-				missingSets[nodeName] = missing
+				// Do sets check.
+				missingSets := map[string][]string{} // node name -> list of missing `sets` args
+				for nodeName, nodeSpec := range seqSpec.Nodes {
+					if nodeSpec.IsJob() {
+						continue
+					}
+					// Compare declared `sets` with what the node
+					// actually sets
+					subseqs := getNodeSubsequences(nodeSpec, gr.sequenceSpecs)
+					sets := getActualSets(subseqs, seqSets)
+					missing := getMissingSets(sets, nodeSpec.Sets)
+					if len(missing) != 0 {
+						missingSets[nodeName] = missing
+					}
+				}
+				if len(missingSets) > 0 {
+					msg := []string{}
+					for nodeName, missing := range missingSets {
+						msg = append(msg, fmt.Sprintf("%s (failed to set %s)", nodeName, strings.Join(missing, ", ")))
+					}
+					multiple := ""
+					if len(missingSets) > 1 {
+						multiple = "s"
+					}
+					seqErrors[seqName] = fmt.Errorf("node%s did not set job args declared in 'sets': %s", multiple, strings.Join(msg, "; "))
+				}
 			}
 		}
 
-		if len(missingSets) > 0 {
-			msg := []string{}
-			for nodeName, missing := range missingSets {
-				msg = append(msg, fmt.Sprintf("%s (failed to set %s)", nodeName, strings.Join(missing, ", ")))
-			}
-			multiple := ""
-			if len(missingSets) > 1 {
-				multiple = "s"
+		// If we were unable to check any sequences, there must be a
+		// cyclical dependency.
+		if !newSeqChecked {
+			for seqName, _ := range seqsToCheck {
+				seqErrors[seqName] = fmt.Errorf("part of cyclical dependency among sequences")
```

Comment:
Does it matter if this ends up overwriting an existing error?

Reply:
Actually, I hadn't noticed that it'd do that. As it is, this will only overwrite errors generated while building the sequence (with `buildSeqGraph`). I'm inclined to preserve the build error. What do you think?

Reply:
I agree it'd be nice to keep the build error somehow, but I think all the sequences that are part of the cyclical dependency issue need to have this error, else it can get confusing.

eg. say there's a cycle of sequences A -> B -> C -> A, but C also had a build error. If we prioritize build > cycle error, than you'll see something like A = art of cyclical dependency, B = art of cyclical dependency, C = build error, which implies that C _isn't_ part of the cycle issue.

Is there any way to preserve both errors? If not, I think overwriting the build error is, while not great, the better option.

Reply:
I could always concatenate the error strings, but that's not particularly pretty. Or possibly make `seqErrors` a `map[string][]string`, but that seems even worse.

In any case, I agree that prioritizing the cycle error makes sense.

Reply:
Yeah, I guess let's just go with the cycle error for now.  `map[string][]string` would work, but it is... ugly, and it's not that bad to just paper over the other error while the cycle error exists.